### ISSUE TITLE
SC2: read in replays and base-train based on them

### ISSUE THREAD

**Issue by espenhk** [espenhk/gamer-ai#349]
> **Original request**
>
> Write a module for SC2 that lets you read in a folder containing replay files, learning from one of the players. The aim is to build weights such that an agent can replicate that behavior, and be further fine-trained from that starting point. Also, look for a database of top-level StarCraft play, likely the same one AlphaStar used if available.

---

## Epic: SC2 replay ingestion → behaviour cloning → fine-tune with any policy

This is the umbrella/epic issue. Implementation is broken into the sub-issues below — **outline only, nothing implemented yet**.

### Overall plan

**Goal.** Read a folder of `.SC2Replay` files, learn from one chosen player (offline behaviour cloning), produce policy weights that (a) replicate that player and (b) serve as a warm-start for **any** SC2-compatible policy to be further RL-fine-tuned.

**How it slots into the existing code** (verified):
- `SC2Client._timestep_to_obs_info` (`games/sc2/client.py:742`) already turns a PySC2 `TimeStep` into the framework obs vector — reused for the obs side of each demonstration pair.
- `_get_pysc2_id_to_fn_idx()` (`client.py:213`) + `FUNCTION_IDS` (`actions.py:73`) give us the PySC2-id → framework-`fn_idx` map; combined with PySC2's `features.reverse_action`, we can invert a replay action into the `[fn_idx, x, y, queue]` target (the inverse of `action_to_function_call`, `actions.py:479`).
- BC writes a `policy_weights.yaml` into the standard experiment dir (`SC2Adapter.experiment_dir`, `adapter.py:82`); the existing SC2 loop auto-loads it, so "fine-tune" is just a normal `python main.py --game sc2` run — mirroring the existing `do_pretrain` precedent (`framework/training.py:1763`).

**Dataset (second ask — answered in [3/6]).** The "AlphaStar dataset" is Blizzard's official SC2 ladder replay packs from `s2client-proto` (`download_replays.py`), gated behind the Blizzard AI & ML License; DeepMind's AlphaStar Unplugged consumes the same packs. We read raw `.SC2Replay` via PySC2 (no TF/tfrecord dependency) and never vendor replays (license + size).

**Proposed workflow once shipped:**
```bash
# 1. base-train (behaviour-clone) from a replay folder — winner of each game, Terran-only
python main.py myrun --game sc2 --bc --replay-dir /path/to/replays --bc-race terran
# 2. fine-tune from those weights with whatever policy_type is configured
python main.py myrun --game sc2
```

### Sub-issues
- [ ] #350 — [1/6] action inversion + shared observation extraction (foundation)
- [ ] #351 — [2/6] replay reader → **sequence-aware** demonstration dataset (winner default + race filter)
- [ ] #352 — [3/6] dataset sourcing (Blizzard / AlphaStar packs) + docs
- [ ] #353 — [4/6] core BC fit + `--bc` entry point (default = MLP)
- [ ] #354 — [5/6] policy-agnostic warm-start — **fine-tune from BC with ANY policy**
- [ ] #355 — [6/6] docs, changelog & test-suite README

### Key design decisions (all settled)
- **Policy-agnostic, sequence-aware dataset.** The reader emits one `demos.npz`: `obs` `[N, obs_dim]`, `actions` `[N, 4]` stored in temporal order, `episode_starts`/`episode_id` delimiting per-game sequences, + JSON `meta`. A loader exposes both a flat per-step view (memoryless policies) and an ordered per-episode view (recurrent policies) — no re-extraction. Sequence-awareness is required, not deferred.
- **Default BC target = MLP** (`sc2_reinforce` two-head: CE on fn + MSE on spatial, gradient pre-train). Linear (`sc2_genetic`) and all other families reachable via `--bc-target` ([#354]).
- **Default player = winner** (resolved via `controller.replay_info`); explicit `1`/`2` allowed.
- **Race filtering of the replay set** — `--bc-race <race>` keeps only replays whose cloned (winning) player is that race, so a mixed folder can base-train a single-race agent (e.g. Terran). Defaults to `any`.
- **6-issue split kept** as listed above.
- **Both CLI flags and `bc_*` config keys** are exposed; CLI overrides `training_params.yaml`.
- SB3 policies are gated off SC2 and out of scope.

_All open questions resolved — plan is ready to implement, starting from #350._
